### PR TITLE
Rust side of password change

### PR DESCRIPTION
**Pull request by bitfriend**
Will fix [this issue](https://github.com/acterglobal/a3/issues/1526), but this PR will not cover password reset via email, because that feature in matrix sdk doesn't seem to be ready.
This PR was separated from [Rust api for password change](https://github.com/acterglobal/a3/pull/1638).

### CODE REVIEW

File: native/acter/src/api/auth.rs [hunk 1]
```diff
@@ -304,23 +305,29 @@ pub async fn register_under_config(
...
+            // user registration follows UIAA spec
+            // req-resp exchange continues until final stage of auth process
+            let request = register::v3::Request::new();
+            if let Err(e) = client.matrix_auth().register(request).await {
+                let Some(inf) = e.as_uiaa_response() else {
+                    bail!("No a uiaa response")
+                };
+                if let Some(err) = &inf.auth_error {
+                    bail!("Found auth error: {:?}", err.message);
+                }
```
Comment (gnunicorn) [inline native/acter/src/api/auth.rs:317]: please note that this has recently be changed to support better error messages to users, please update the code here accordingly.

File: native/acter/src/api/auth.rs [hunk 2]
```diff
@@ -304,23 +305,29 @@ pub async fn register_under_config(
...
+                let has_flow = inf
+                    .flows
+                    .iter()
+                    .any(|f| f.stages.contains(&AuthType::Dummy));
+                let was_completed = inf.completed.contains(&AuthType::Dummy);
+                if has_flow && !was_completed {
```
Comment (gnunicorn) [inline native/acter/src/api/auth.rs:323]: I don't understand what this check is for ... if it doesn't have the flow we are just not continuing? not even showing an error? that seems wrong.

File: native/acter/src/api/auth.rs [hunk 3]
```diff
@@ -387,29 +394,35 @@ pub async fn register_with_token_under_config(
...
+                    let reg_token = assign!(RegistrationToken::new(registration_token), {
+                        session: inf.session.clone(),
+                    });
+                    let request = assign!(register::v3::Request::new(), {
+                        auth: Some(AuthData::RegistrationToken(reg_token)),
+                    });
+                    client.matrix_auth().register(request).await?;
```
Comment (gnunicorn) [inline native/acter/src/api/auth.rs:424]: same remarks as above.

File: native/acter/src/api/auth.rs [hunk 4]
```diff
@@ -432,15 +445,50 @@ pub async fn register_with_token_under_config(
...
+                        bail!("Found auth error: {:?}", err.message);
+                    }
+                    let pswd = assign!(Password::new(account.user_id().into(), password), {
+                        session: inf.session.clone(),
+                    });
+                    let auth_data = AuthData::Password(pswd);
+                    account.deactivate(None, Some(auth_data)).await?;
+                    // FIXME: remove local data, too!
+                }
+                Ok(true)
```
Comment (gnunicorn) [inline native/acter/src/api/auth.rs:465]: I don't understand this change either...

File: native/acter/src/api/auth.rs [hunk 5]
```diff
@@ -432,15 +445,50 @@ pub async fn register_with_token_under_config(
...
+    }
+
+    pub async fn change_password(&self, old_val: String, new_val: String) -> Result<bool> {
+        let client = self.core.client().clone();
+        let account = self.account()?;
+        RUNTIME
+            .spawn(async move {
+                let capabilities = client.get_capabilities().await?;
+                if !capabilities.change_password.enabled {
+                    bail!("This client doesn't support password change");
```
Comment (gnunicorn) [inline native/acter/src/api/auth.rs:477]:
```suggestion
                    bail!("Server doesn't support password change");
```

File: native/acter/src/api/auth.rs [hunk 6]
```diff
@@ -432,15 +445,50 @@ pub async fn register_with_token_under_config(
...
+                }
+                if let Err(e) = account.change_password(&new_val, None).await {
+                    let Some(inf) = e.as_uiaa_response() else {
+                        bail!("Server did not indicate how to allow password change.")
+                    };
+                    if let Some(err) = &inf.auth_error {
+                        bail!("Found auth error: {:?}", err.message);
+                    }
```
Comment (gnunicorn) [inline native/acter/src/api/auth.rs:485]: please use the new facilities to handle the error to show more understandable user errors.

**Review by gnunicorn — CHANGES_REQUESTED** [review]
Password change code seems fine, other than a small update on handling errors that has been introduced recently. If it was only that, it would get an approval. But there's a few unrelated and unexplained changes on the registration flows as well, I don't understand the need for and bluntly think are wrong. This can't be merged as such. I'd recommend separating out the specific password-change part and resubmit that as a separate, clean PR instead.

File: native/acter/src/api/auth.rs (hunk 2 above)
Comment (bitfriend) [reply native/acter/src/api/auth.rs]: Fixed by https://github.com/acterglobal/a3/pull/1761/commits/7c3219bcea8563bf8049f8be453e1fc735db229e

File: native/acter/src/api/auth.rs (hunk 3 above)
Comment (bitfriend) [reply native/acter/src/api/auth.rs]: Fixed by https://github.com/acterglobal/a3/pull/1761/commits/7c3219bcea8563bf8049f8be453e1fc735db229e

File: native/acter/src/api/auth.rs (hunk 5 above)
Comment (bitfriend) [reply native/acter/src/api/auth.rs]: Fixed by https://github.com/acterglobal/a3/pull/1761/commits/674119aa70a60b3d3efc30ec82122fe34c2703d4

File: native/acter/src/api/auth.rs (hunk 1 above)
Comment (bitfriend) [reply native/acter/src/api/auth.rs]: Fixed by https://github.com/acterglobal/a3/pull/1761/commits/4f1ee0e0417fc7e1ecc0119ff65aa79e6c6f8461

File: native/acter/src/api/auth.rs (hunk 6 above)
Comment (bitfriend) [reply native/acter/src/api/auth.rs]: Fixed by https://github.com/acterglobal/a3/pull/1761/commits/4f1ee0e0417fc7e1ecc0119ff65aa79e6c6f8461

**Review by gnunicorn — APPROVED** [review]
LGTM

**Pull request merged**